commit fe20afaee8aac7838ed6e4a76baa83e547629582
Author: Tom Lane <tgl@sss.pgh.pa.us>
Date:   Mon May 9 14:15:37 2022 -0400

    Fix core dump in transformValuesClause when there are no columns.
    
    The parser code that transformed VALUES from row-oriented to
    column-oriented lists failed if there were zero columns.
    You can't write that straightforwardly (though probably you
    should be able to), but the case can be reached by expanding
    a "tab.*" reference to a zero-column table.
    
    Per bug #17477 from Wang Ke.  Back-patch to all supported branches.
    
    Discussion: https://postgr.es/m/17477-0af3c6ac6b0a6ae0@postgresql.org

diff --git a/src/backend/parser/analyze.c b/src/backend/parser/analyze.c
index 0144284aa35..6b54e8e46df 100644
--- a/src/backend/parser/analyze.c
+++ b/src/backend/parser/analyze.c
@@ -1422,11 +1422,11 @@ transformSelectStmt(ParseState *pstate, SelectStmt *stmt)
  */
 static Query *
 transformValuesClause(ParseState *pstate, SelectStmt *stmt)
 {
 	Query	   *qry = makeNode(Query);
-	List	   *exprsLists;
+	List	   *exprsLists = NIL;
 	List	   *coltypes = NIL;
 	List	   *coltypmods = NIL;
 	List	   *colcollations = NIL;
 	List	  **colexprs = NULL;
 	int			sublist_length = -1;
@@ -1506,10 +1506,13 @@ transformValuesClause(ParseState *pstate, SelectStmt *stmt)
 			i++;
 		}
 
 		/* Release sub-list's cells to save memory */
 		list_free(sublist);
+
+		/* Prepare an exprsLists element for this row */
+		exprsLists = lappend(exprsLists, NIL);
 	}
 
 	/*
 	 * Now resolve the common types of the columns, and coerce everything to
 	 * those types.  Then identify the common typmod and common collation, if
@@ -1549,28 +1552,19 @@ transformValuesClause(ParseState *pstate, SelectStmt *stmt)
 	}
 
 	/*
 	 * Finally, rearrange the coerced expressions into row-organized lists.
 	 */
-	exprsLists = NIL;
-	foreach(lc, colexprs[0])
-	{
-		Node	   *col = (Node *) lfirst(lc);
-		List	   *sublist;
-
-		sublist = list_make1(col);
-		exprsLists = lappend(exprsLists, sublist);
-	}
-	list_free(colexprs[0]);
-	for (i = 1; i < sublist_length; i++)
+	for (i = 0; i < sublist_length; i++)
 	{
 		forboth(lc, colexprs[i], lc2, exprsLists)
 		{
 			Node	   *col = (Node *) lfirst(lc);
 			List	   *sublist = lfirst(lc2);
 
 			sublist = lappend(sublist, col);
+			lfirst(lc2) = sublist;
 		}
 		list_free(colexprs[i]);
 	}
 
 	/*
diff --git a/src/test/regress/expected/select.out b/src/test/regress/expected/select.out
index d7888e44eba..33a6dceb0e3 100644
--- a/src/test/regress/expected/select.out
+++ b/src/test/regress/expected/select.out
@@ -512,10 +512,17 @@ TABLE int8_tbl;
  4567890123456789 |               123
  4567890123456789 |  4567890123456789
  4567890123456789 | -4567890123456789
 (9 rows)
 
+-- corner case: VALUES with no columns
+CREATE TEMP TABLE nocols();
+INSERT INTO nocols DEFAULT VALUES;
+SELECT * FROM nocols n, LATERAL (VALUES(n.*)) v;
+--
+(1 row)
+
 --
 -- Test ORDER BY options
 --
 CREATE TEMP TABLE foo (f1 int);
 INSERT INTO foo VALUES (42),(3),(10),(7),(null),(null),(1);
diff --git a/src/test/regress/sql/select.sql b/src/test/regress/sql/select.sql
index aa3d0352046..019f1e76739 100644
--- a/src/test/regress/sql/select.sql
+++ b/src/test/regress/sql/select.sql
@@ -141,10 +141,15 @@ VALUES (1,2), (3,4+4), (7,77.7)
 UNION ALL
 SELECT 2+2, 57
 UNION ALL
 TABLE int8_tbl;
 
+-- corner case: VALUES with no columns
+CREATE TEMP TABLE nocols();
+INSERT INTO nocols DEFAULT VALUES;
+SELECT * FROM nocols n, LATERAL (VALUES(n.*)) v;
+
 --
 -- Test ORDER BY options
 --
 
 CREATE TEMP TABLE foo (f1 int);
